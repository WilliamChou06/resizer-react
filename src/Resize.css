body,
html {
  background: black;
}
.resizable {
  background: white;
  width: 100px;
  height: 100px;
  position: absolute;
  top: 100px;
  left: 100px;
}

.resizable .resizers{
  width: 100%;
  height: 100%;
  border: 3px solid #4286f4;
  box-sizing: border-box;
}

.resizable .resizer{
  background: white;
  border: 3px solid #4286f4;
  position: absolute;
}

.resizable .resizer.top-left {
  left: 0;
  top: 0;
  cursor: nwse-resize; /*resizer cursor*/
}
.resizable .resizer.top-right {
  right: 0;
  top: 0;
  cursor: nesw-resize;
}
.resizable .resizer.bottom-left {
  left: 0;
  bottom: 0;
  cursor: nesw-resize;
}
.resizable .resizer.bottom-right {
  right: 0;
  bottom: 0;
  cursor: nwse-resize;
}

#top {
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  box-sizing: border-box;
  cursor: ns-resize;
}
#right {
  top: 0;
  right: 0;
  bottom: 0;
  width: 5px;
  box-sizing: border-box;
  cursor: ew-resize;
}
#bottom {
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  box-sizing: border-box;
  cursor: ns-resize;
}
#left {
  top: 0;
  left: 0;
  bottom: 0;
  width: 5px;
  box-sizing: border-box;
  cursor: ew-resize;
}