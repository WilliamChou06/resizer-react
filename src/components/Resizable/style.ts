import styled from 'styled-components';

export const StyledResizableDiv = styled.div`
  width: ${props => props.width || 120} !important;
  height: ${props => props.height || 120} !important;
  left: ${props => props.left || 0} !important;
  top: ${props => props.top || 0} !important;
  background: ${props => props.background || 'white'} !important;
  position: absolute;
  user-select: none;
  overflow: hidden;
  -moz-user-select: none;
  -khtml-user-select: none;
  -webkit-user-select: none;
  -o-user-select: none;

  .handle {
    background: ${props => props.background || 'white'} !important;
  }
`;
