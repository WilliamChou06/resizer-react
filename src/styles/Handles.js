import styled, { css } from 'styled-components';

const resizer = css`
  background: white;
  border: 3px solid #4286f4;
  position: absolute;
`;

export const TopHandle = styled.div`
  ${resizer}
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  box-sizing: border-box;
  cursor: ns-resize;
`;
export const RightHandle = styled.div`
  ${resizer}
  top: 0;
  right: 0;
  bottom: 0;
  width: 5px;
  box-sizing: border-box;
  cursor: ew-resize;
`;
export const BottomHandle = styled.div`
  ${resizer}
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  box-sizing: border-box;
  cursor: ns-resize;
`;
export const LeftHandle = styled.div`
  ${resizer}
  top: 0;
  left: 0;
  bottom: 0;
  width: 5px;
  box-sizing: border-box;
  cursor: ew-resize;
`;
export const TopLeftHandle = styled.div`
  ${resizer}
  left: 0;
  top: 0;
  width: 5px;
  cursor: nwse-resize;
`;
export const TopRightHandle = styled.div`
  ${resizer}
  right: 0;
  top: 0;
  width: 5px;

  cursor: nesw-resize;
`;
export const BottomLeftHandle = styled.div`
  ${resizer}
  left: 0;
  bottom: 0;
  width: 5px;
  cursor: nesw-resize;
`;
export const BottomRightHandle = styled.div`
  ${resizer}
  right: 0;
  bottom: 0;
  width: 5px;
  cursor: nwse-resize;
`;
