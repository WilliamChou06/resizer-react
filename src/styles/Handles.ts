import styled, { css } from 'styled-components';

const resizer = css`
  background: white;
  border: 3px solid transparent;
  position: absolute;
`;

const smWidth = css`
  width: 5px
`

export const TopHandle = styled.div`
  ${resizer}
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  box-sizing: border-box;
  cursor: ns-resize;
`;
export const RightHandle = styled.div`
  ${resizer}
  ${smWidth}
  top: 0;
  right: 0;
  bottom: 0;
  width: 5px;
  box-sizing: border-box;
  cursor: ew-resize;
`;
export const BottomHandle = styled.div`
  ${resizer}
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  box-sizing: border-box;
  cursor: ns-resize;
`;
export const LeftHandle = styled.div`
  ${resizer}
  ${smWidth}
  top: 0;
  left: 0;
  bottom: 0;
  box-sizing: border-box;
  cursor: ew-resize;
`;
export const TopLeftHandle = styled.div`
  ${resizer}
  ${smWidth}
  left: 0;
  top: 0;
  cursor: nwse-resize;
`;
export const TopRightHandle = styled.div`
  ${resizer}
  ${smWidth}
  right: 0;
  top: 0;
  cursor: nesw-resize;
`;
export const BottomLeftHandle = styled.div`
  ${resizer}
  ${smWidth}
  left: 0;
  bottom: 0;
  cursor: nesw-resize;
`;
export const BottomRightHandle = styled.div`
  ${resizer}
  ${smWidth}
  right: 0;
  bottom: 0;
  cursor: nwse-resize;
`;
