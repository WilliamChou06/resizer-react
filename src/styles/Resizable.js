import styled from 'styled-components';

export const StyledResizableDiv = styled.div`
  width: ${props => props.width || 120} !important;
  height: ${props => props.height || 120} !important;
  left: ${props => props.left || 0} !important;
  top: ${props => props.top || 0} !important;
  background: white;
  height: 100px;
  position: absolute;
  top: 100px;
  left: 100px;
  user-select: none;
  -moz-user-select: none;
  -khtml-user-select: none;
  -webkit-user-select: none;
  -o-user-select: none;
`;

export const ResizableContent = styled.div`
  word-wrap: break-word;
`;
