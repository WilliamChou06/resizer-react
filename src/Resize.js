import React from 'react';
import './Resize.css';

function Resizable(props) {
  return (
    <div className="resizable">
      <div className="resizers">
        <div className="resizer top-left" />
        <div className="resizer top-right" />
        <div className="resizer bottom-left" />
        <div className="resizer bottom-right" />
        {props.children}
      </div>
    </div>
  );
}

export default Resizable;
