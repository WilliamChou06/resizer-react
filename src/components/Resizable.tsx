import React, { Component } from 'react';
import { StyledResizableDiv } from '../styles/Resizable';
import ResizingHandles from './ResizingHandles';

interface Props {
  width?: number,
  height?: number
}
interface State {
  style: Style
  resizing: string,
  bounding: object,
  mouseLocation: MouseLocation
  originalBounding: object,
}

interface Style {
  width: number,
  height: number,
  top: number,
  left: number
  originalWidth: number,
  originalHeight: number,
}

interface MouseLocation {
  mouseX: number,
  mouseY: number,
}

class Resizable extends Component<Props, State> {
  state = {
    bounding: null,
    originalBounding: null,
    style: {
      width: this.props.width || null,
      height: this.props.height || null,
      top: null,
      left: null,
      originalWidth: this.props.width,
      originalHeight: this.props.height,
    },
    resizing: '',
    mouseLocation: {
      mouseX: 0,
      mouseY: 0,
    },
  };

  // Use events on windows so it doesn't stop working when dragged outside the resizable component
  componentDidMount() {
    window.addEventListener('mouseup', this.handleMouseUp);
  }

  componentWillUnmount() {
    window.removeEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseDown = (e: any): void => {
    // Setting starting values to calculate next values.
    this.setState({
      mouseLocation: {
        ...this.state.mouseLocation, mouseX: e.pageX,
        mouseY: e.pageY,
      },
      style: { ...this.state.style, originalHeight: this.state.style.height, originalWidth: this.state.style.width },
      originalBounding: this.state.bounding
    });

    console.log(this.state.style)

    // Set state to which handle is being dragged

    switch (e.target.id) {
      case 'top-left':
        this.setState({ resizing: 'top-left' });
        break;
      case 'top':
        this.setState({ resizing: 'top' });
        break;
      case 'top-right':
        this.setState({ resizing: 'top-right' });
        break;
      case 'right':
        this.setState({ resizing: 'right' });
        break;
      case 'bottom-left':
        this.setState({ resizing: 'bottom-left' });
        break;
      case 'bottom':
        this.setState({ resizing: 'bottom' });
        break;
      case 'bottom-right':
        this.setState({ resizing: 'bottom-right' });
        break;
      case 'left':
        this.setState({ resizing: 'left' });
        break;
      default:
        this.setState({ resizing: '' });
        break;
    }

    // Event listener to window so it doesn't stop resizing if dragged outside of resizable component
    window.addEventListener('mousemove', this.handleMouseMove);
  };

  handleMouseMove = (e: MouseEvent): void => {
    let originalMouseX = this.state.mouseLocation.mouseX;
    let originalMouseY = this.state.mouseLocation.mouseY;
    let originalWidth = this.state.style.originalWidth;
    let originalHeight = this.state.style.originalHeight;
    let bounding = this.state.bounding;

    // Check for handle being dragged and calculate width and height accordingly

    switch (this.state.resizing) {
      case 'top-left':
        this.setState({
          style: {
            ...this.state.style, width: originalWidth - (e.pageX - originalMouseX),
            height: originalHeight - (e.pageY - originalMouseY),
            left: bounding.left + (e.pageX - bounding.left),
            top: bounding.top + (e.pageY - bounding.top)
          }
        });
        break;
      case 'top':
        this.setState({
          style: {
            ...this.state.style, height: originalHeight - (e.pageY - originalMouseY),
            top: bounding.top + (e.pageY - bounding.top)
          }
        });
        break;
      case 'top-right':
        this.setState({
          style: {
            ...this.state.style, width: originalWidth + (e.pageX - originalMouseX),
            height: originalHeight - (e.pageY - originalMouseY),
            top: bounding.top + (e.pageY - bounding.top)
          }

        });
        break;
      case 'right':
        this.setState({
          style: {
            ...this.state.style, width: originalWidth + (e.pageX - originalMouseX)
          }

        });
        break;
      case 'bottom-left':
        this.setState({
          style: {
            ...this.state.style, width: originalWidth - (e.pageX - originalMouseX),
            height: originalHeight + (e.pageY - originalMouseY),
            left: bounding.left + (e.pageX - bounding.left)
          }


        });
        break;
      case 'bottom':
        this.setState({
          style: {
            ...this.state.style, height: originalHeight + (e.pageY - originalMouseY)
          }

        });
        break;
      case 'bottom-right':
        this.setState({
          style: {
            ...this.state.style, width: originalWidth + (e.pageX - originalMouseX),
            height: e.pageY - bounding.top
          }


        });
        break;
      case 'left':
        this.setState({
          style: {
            ...this.state.style,
            width: originalWidth - (e.pageX - originalMouseX),
            left: bounding.left + (e.pageX - bounding.left)
          }
        });
        break;
      default:
        break;
    }
  };

  handleMouseUp = (): void => {
    this.setState({ resizing: '' });
    window.removeEventListener('mousemove', this.handleMouseMove);
  };

  getBoundingClientRect = (el: HTMLElement): void => {

    // getBoundingClientRect using refs
    if (el) {
      console.log(el);
      this.setState({ bounding: el.getBoundingClientRect() });
      if (!this.props.width && !this.props.height) {
        // this.setState({ height: el.parentElement.clientHeight, width: el.parentElement.clientWidth })
        this.setState({ style: { ...this.state.style, height: el.parentElement.clientHeight, width: el.parentElement.clientWidth } })
      }
    }
  };

  render() {
    return (
      <StyledResizableDiv
        onMouseDown={this.handleMouseDown}
        ref={this.getBoundingClientRect}
        width={this.state.style.width && this.state.style.width + 'px'}
        height={this.state.style.height && this.state.style.height + 'px'}
        top={this.state.style.top + 'px'}
        left={this.state.style.left + 'px'}
      >
        <ResizingHandles>{this.props.children}</ResizingHandles>
      </StyledResizableDiv>
    );
  }
}

export default Resizable;
