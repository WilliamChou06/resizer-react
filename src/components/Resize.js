import React, { Component } from 'react';
import {
  TopHandle,
  TopLeftHandle,
  TopRightHandle,
  RightHandle,
  BottomRightHandle,
  BottomHandle,
  BottomLeftHandle,
  LeftHandle
} from '../styles/Handles';
import { ResizableDiv, ResizableContent } from '../styles/Resizable';

class Resizable extends Component {
  state = {
    width: this.props.width || 120,
    height: this.props.height || 120,
    resizing: true,
    bounding: 0,
    mouseX: 0,
    mouseY: 0,
    originalWidth: 120,
    originalHeight: 120,
    originalBounding: 0
  };

  // Use events on windows so it doesn't stop working when dragged outside the resizable component

  componentDidMount() {
    window.addEventListener('mouseup', this.handleMouseUp);
  }

  componentWillUnmount() {
    window.removeEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseDown = e => {
    // Setting starting values to calculate next values.

    this.setState({
      mouseX: e.pageX,
      mouseY: e.pageY,
      originalHeight: this.state.height,
      originalWidth: this.state.width,
      originalBounding: this.state.bounding
    });

    // Set state to which handle is being dragged

    switch (e.target.id) {
      case 'top-left':
        this.setState({ resizing: 'top-left' });
        break;
      case 'top':
        this.setState({ resizing: 'top' });
        break;
      case 'top-right':
        this.setState({ resizing: 'top-right' });
        break;
      case 'right':
        this.setState({ resizing: 'right' });
        break;
      case 'bottom-left':
        this.setState({ resizing: 'bottom-left' });
        break;
      case 'bottom':
        this.setState({ resizing: 'bottom' });
        break;
      case 'bottom-right':
        this.setState({ resizing: 'bottom-right' });
        break;
      case 'left':
        this.setState({ resizing: 'left' });
        break;
      default:
        break;
    }

    // Event listener to window so it doesn't stop resizing if dragged outside of resizable component
    window.addEventListener('mousemove', this.handleMouseMove);
  };

  handleMouseMove = e => {
    let originalMouseX = this.state.mouseX;
    let originalMouseY = this.state.mouseY;
    let originalWidth = this.state.originalWidth;
    let originalHeight = this.state.originalHeight;
    let bounding = this.state.bounding;

    // Check for handle being dragged and calculate width and height accordingly

    if (this.state.resizing === 'bottom-right') {
      this.setState({
        width: originalWidth + (e.pageX - originalMouseX),
        height: e.pageY - bounding.top
      });
    }
    if (this.state.resizing === 'bottom-left') {
      this.setState({
        width: originalWidth - (e.pageX - originalMouseX),
        height: originalHeight + (e.pageY - originalMouseY),
        left: bounding.left + (e.pageX - bounding.left)
      });
    }

    if (this.state.resizing === 'top-right') {
      this.setState({
        width: originalWidth + (e.pageX - originalMouseX),
        height: originalHeight - (e.pageY - originalMouseY),
        top: bounding.top + (e.pageY - bounding.top)
      });
    }

    if (this.state.resizing === 'top-left') {
      this.setState({
        width: originalWidth - (e.pageX - originalMouseX),
        height: originalHeight - (e.pageY - originalMouseY),
        left: bounding.left + (e.pageX - bounding.left),
        top: bounding.top + (e.pageY - bounding.top)
      });
    }
    if (this.state.resizing === 'top') {
      this.setState({
        height: originalHeight - (e.pageY - originalMouseY),
        top: bounding.top + (e.pageY - bounding.top)
      });
    }
    if (this.state.resizing === 'right') {
      this.setState({
        width: originalWidth + (e.pageX - originalMouseX)
      });
    }
    if (this.state.resizing === 'bottom') {
      this.setState({
        height: originalHeight + (e.pageY - originalMouseY)
      });
    }
    if (this.state.resizing === 'left') {
      this.setState({
        width: originalWidth - (e.pageX - originalMouseX),
        left: bounding.left + (e.pageX - bounding.left)
      });
    }
  };

  handleMouseUp = () => {
    this.setState({ resizing: false });
    window.removeEventListener('mousemove', this.handleMouseMove);
  };

  getBoundingClientRect = el => {
    // getBoundingClientRect using refs

    if (el) {
      this.setState({ bounding: el.getBoundingClientRect() });
    }
  };

  render() {
    return (
      <ResizableDiv
        onMouseDown={this.handleMouseDown}
        ref={this.getBoundingClientRect}
        width={this.state.width + 'px'}
        height={this.state.height + 'px'}
        top={this.state.top + 'px'}
        left={this.state.left + 'px'}
      >
        <div>
          <TopHandle id="top" />
          <RightHandle id="right" />
          <BottomHandle id="bottom" />
          <LeftHandle id="left" />
          <TopLeftHandle id="top-left" />
          <TopRightHandle id="top-right" />
          <BottomLeftHandle id="bottom-left" />
          <BottomRightHandle id="bottom-right" />
          {/* Logging */}
          <ResizableContent>{this.props.children}</ResizableContent>
          {this.state.width}
          {this.state.height}
          {this.state.bounding.left}
        </div>
      </ResizableDiv>
    );
  }
}

export default Resizable;
