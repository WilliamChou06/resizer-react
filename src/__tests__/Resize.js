import React from 'react';
import { shallow } from 'enzyme';
import Resize from '../components/Resize';
import { TopHandle } from '../styles/Handles';
import { ResizableDiv } from '../styles/Resizable';

let wrapper, handleMouseDown;

beforeEach(() => {
  handleMouseDown = jest.fn();
  wrapper = shallow(<Resize />);
});

it('renders Resize component', () => {
  expect(wrapper).toMatchSnapshot();
});

// it('handles handle mousedown', () => {
//   wrapper.find(ResizableDiv).simulate('mousedown');
//   expect(handleMouseDown).toHaveBeenCalled();
// });
