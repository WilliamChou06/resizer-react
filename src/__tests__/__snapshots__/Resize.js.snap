// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Resize component 1`] = `ShallowWrapper {}`;
